src/gfx/pal_packing.cpp

```diff
@@ -0,0 +1,359 @@
+/*
+ * This file is part of RGBDS.
+ *
+ * Copyright (c) 2022, Eldred Habert and RGBDS contributors.
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include "gfx/pal_packing.hpp"
+
+#include <assert.h>
+#include <bitset>
+#include <inttypes.h>
+#include <numeric>
+#include <optional>
+#include <queue>
+#include <tuple>
+#include <type_traits>
+#include <unordered_set>
+#include <vector>
+
+#include "gfx/main.hpp"
+#include "gfx/proto_palette.hpp"
+
+using std::swap;
+
+namespace packing {
+
+// The solvers here are picked from the paper at http://arxiv.org/abs/1605.00558:
+// "Algorithms for the Pagination Problem, a Bin Packing with Overlapping Items"
+// Their formulation of the problem consists in packing "tiles" into "pages"; here is a
+// correspondence table for our application of it:
+// Paper | RGBGFX
+// ------+-------
+//  Tile | Proto-palette
+//  Page | Palette
+
+/**
+ * A reference to a proto-palette, and attached attributes for sorting purposes
+ */
+struct ProtoPalAttrs {
+	size_t const palIndex;
+	/**
+	 * Pages from which we are banned (to prevent infinite loops)
+	 * This is dynamic because we wish not to hard-cap the amount of palettes
+	 */
+	std::vector<bool> bannedPages;
+
+	ProtoPalAttrs(size_t index) : palIndex(index) {}
+	bool isBannedFrom(size_t index) const {
+		return index < bannedPages.size() && bannedPages[index];
+	}
+	void banFrom(size_t index) {
+		if (bannedPages.size() <= index) {
+			bannedPages.resize(index + 1);
+		}
+		bannedPages[index] = true;
+	}
+};
+
+/**
+ * A collection of proto-palettes assigned to a palette
+ * Does not contain the actual color indices because we need to be able to remove elements
+ */
+class AssignedProtos {
+	// We leave room for emptied slots to avoid copying the structs around on removal
+	std::vector<std::optional<ProtoPalAttrs>> _assigned;
+	// For resolving proto-palette indices
+	std::vector<ProtoPalette> const &_protoPals;
+
+public:
+	template<typename... Ts>
+	AssignedProtos(decltype(_protoPals) protoPals, Ts &&...elems)
+		: _assigned{std::forward<Ts>(elems)...}, _protoPals{protoPals} {}
+
+private:
+	template<typename Inner, template<typename> typename Constness>
+	class Iter {
+	public:
+		friend class AssignedProtos;
+		// For `iterator_traits`
+		using difference_type = typename std::iterator_traits<Inner>::difference_type;
+		using value_type = ProtoPalAttrs;
+		using pointer = Constness<value_type> *;
+		using reference = Constness<value_type> &;
+		using iterator_category = std::input_iterator_tag;
+
+	private:
+		Constness<decltype(_assigned)> *_array = nullptr;
+		Inner _iter{};
+
+		Iter(decltype(_array) array, decltype(_iter) &&iter) : _array(array), _iter(iter) {
+			skipEmpty();
+		}
+		void skipEmpty() {
+			while (_iter != _array->end() && !(*_iter).has_value()) {
```

Comment:
Would this not work?
```suggestion
			while (_iter != _array->end() && !_iter->has_value()) {
```

Reply:
Yep, that's a holdover from an old version.